tokstr.c

```diff
@@ -75,22 +76,49 @@ tokstr_string(const char *source) {
 	return (struct tokstr *) ts;
 }
 
-// tokstr_next -- return next token from an iterator (caller must free() this)
+// tokstr_next -- return next token from an iterator (strndup)
 char *
 tokstr_next(struct tokstr *ts_pub, const char *delims) {
+	struct tokstr_reg reg = tokstr_next_region(ts_pub, delims);
+	if (reg.base == NULL)
+		return NULL;
+	return strndup(reg.base, reg.size);
+}
+
+// tokstr_next_region -- return next token from an iterator (zero-copy)
+struct tokstr_reg
+tokstr_next_region(struct tokstr *ts_pub, const char *delims) {
 	struct tokstr_pvt *ts = (struct tokstr_pvt *) ts_pub;
-	char *ret = NULL;
+	struct tokstr_reg reg = {};
 	switch (ts->data.class.type) {
 	case ts_buffer:
-		ret = next_buffer(&ts->data.buffer, delims);
+		reg = next_region(&ts->data.region, delims);
 		break;
 	case ts_string:
-		ret = next_string(&ts->data.string, delims);
+		reg = next_string(&ts->data.string, delims);
 		break;
 	default:
 		abort();
```

Comment:
The calls to abort() and assert() _should_ never happen, but it's disturbing that another library is potentially calling them.  This bit us in the mtbl or dnstable libraries.

Reply:
abort and assert here are documenting some invariants. if either such does not hold then the program state is undefined, and cannot be recovered by defining and returning an error code -- the caller would still have to exit although it may try to log an error message first (assuming stdio and syslog aren't also smashed.)

```
// tokstr_next_region -- return next token from an iterator (zero-copy)
struct tokstr_reg
tokstr_next_region(struct tokstr *ts_pub, const char *delims) {
        struct tokstr_pvt *ts = (struct tokstr_pvt *) ts_pub;
        struct tokstr_reg reg = {};
        switch (ts->data.class.type) {
        case ts_buffer:
                reg = next_region(&ts->data.region, delims);
                break;
        case ts_string:
                reg = next_string(&ts->data.string, delims);
                break;
        default:
                abort();
        }
        assert((reg.base == NULL) == (reg.size == 0));
        return reg;
}
```

in the first case, an unenumerated value for `ts->data.class.type` indicates either memory corruption or a linker error such that the tokstr.h and tokstr.o files are out of synch. in the second case, a region whose base is NULL but size is nonzero cannot be generated by the (static) next_region() or next_string() functions, so if it is seen, memory corruption is indicated. likewise a region whose base is NOT NULL but whose size is zero.

please suggest an alternative?

Reply:
As you write, it might be necessary, but document it well for the callers.

Reply:
fixed (ddd2a0a..fca0d00).